transformer_engine/pytorch/attention.py

```diff
@@ -489,52 +490,87 @@ def forward(
             query_layer.dtype in [torch.float16, torch.bfloat16]
             and key_layer.dtype in [torch.float16, torch.bfloat16]
             and value_layer.dtype in [torch.float16, torch.bfloat16]
-            ), 'FlashAttention currently only supports FP16 and BF16.'
+            ), "FlashAttention currently only supports FP16 and BF16."
         assert (
             query_layer.is_cuda and key_layer.is_cuda and value_layer.is_cuda
-            ), 'FlashAttention currently only supports CUDA tensors.'
-
-        # For now just 128, will make it more general in the future
-
-        if (query_layer.shape[-1] == 128 and
-            query_layer.shape[0] * query_layer.shape[1] >= 512 and
-            _check_qkv_layout(query_layer, key_layer, value_layer) == "sbh3d"):
-            query_layer, key_layer, value_layer = _PrepareQKVForFA.apply(query_layer,
-                                                                         key_layer,
-                                                                         value_layer)
-        else:
-            query_layer, key_layer, value_layer = [x.transpose(0,1).contiguous()
-                           for x in (query_layer, key_layer, value_layer)]
-
-        batch_size, seqlen = query_layer.shape[0], query_layer.shape[1]
-
-        # [b, sq, np, hn]
-        query_layer, key_layer, value_layer = [
-            x.view(x.shape[0] * x.shape[1], *x.shape[2:])
-            for x in [query_layer, key_layer, value_layer]
-        ]
-
-        max_seqlen = seqlen
-        cu_seqlens = torch.arange(
-            0,
-            (batch_size + 1) * seqlen,
-            step=seqlen,
-            dtype=torch.int32,
-            device=query_layer.device)
+            ), "FlashAttention currently only supports CUDA tensors."
+        assert (
+            qkv_layout in QKVLayouts
+            ), f"FlashAttention does not support qkv_layout = {qkv_layout}!"
+
+        qkv_format = ''.join([i for i in qkv_layout.split('_')[0] if i.isalpha()])
+
+        if qkv_format == 'sbhd':
+            # For now just 128, will make it more general in the future
+            if (query_layer.shape[-1] == 128 and
+                query_layer.shape[0] * query_layer.shape[1] >= 512 and
+                qkv_layout == "sbh3d"):
+                query_layer, key_layer, value_layer = _PrepareQKVForFA.apply(query_layer,
+                                                                             key_layer,
+                                                                             value_layer)
+            else:
+                query_layer, key_layer, value_layer = [x.contiguous()
+                    for x in (query_layer, key_layer, value_layer)]
+                query_layer, key_layer, value_layer = [x.transpose(0,1).contiguous()
+                    for x in (query_layer, key_layer, value_layer)]
```

Comment:
This seems redundant:
```suggestion
                query_layer, key_layer, value_layer = [x.transpose(0,1).contiguous()
                    for x in (query_layer, key_layer, value_layer)]
```

Reply:
Done, thanks!